deb/local.go

```diff
@@ -175,26 +186,39 @@ func (collection *LocalRepoCollection) LoadComplete(repo *LocalRepo) error {
 
 // ByName looks up repository by name
 func (collection *LocalRepoCollection) ByName(name string) (*LocalRepo, error) {
-	collection.loadList()
-
-	for _, r := range collection.list {
-		if r.Name == name {
-			return r, nil
-		}
+	result := collection.search(func(r *LocalRepo) bool { return r.Name == name }, true)
+	if len(result) == 0 {
+		return nil, fmt.Errorf("local repo with name %s not found", name)
 	}
-	return nil, fmt.Errorf("local repo with name %s not found", name)
+
+	return result[0], nil
 }
 
 // ByUUID looks up repository by uuid
 func (collection *LocalRepoCollection) ByUUID(uuid string) (*LocalRepo, error) {
-	collection.loadList()
+	if r, ok := collection.cache[uuid]; ok {
+		return r, nil
+	}
 
-	for _, r := range collection.list {
-		if r.UUID == uuid {
-			return r, nil
-		}
+	key := (&LocalRepo{UUID: uuid}).Key()
+
+	value, err := collection.db.Get(key)
+	if err == database.ErrNotFound {
+		return nil, fmt.Errorf("local repo with uuid %s not found", uuid)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	r := &LocalRepo{}
```

Comment:
this doesn't seem to be covered by a test (https://codecov.io/gh/aptly-dev/aptly/pull/766/diff#D2-210)

I think as this is a success path it would be good to have a unit test for it. There are some similar code bits uncovered in the other collections too.

Reply:
yep, thanks, I will check it!

Reply:
I've improved test coverage and added some benchmarks for `ByUUID()` and `ByName()`. They exercise "worst case": collection is created and one lookup is performed. On `master` branch this always leads to loading all the objects. 

For `ByUUID()` in the new approach loading is fast, as object is looked up directly (loading only single element). `ByName()` still requires scanning whole collection, in the worst case it is as bad as the old approach, on average it's 50% better.

`ForEach()` doesn't cache objects in memory, so this should help #761 as objects would be GCed as soon as they're scanned. `ByUUID()` is used a lot to lookup source of published repositories.

On `master`:

```
BenchmarkSnapshotCollectionByUUID-8    	     500	   2953932 ns/op	 1352168 B/op	   30743 allocs/op
BenchmarkSnapshotCollectionByName-8    	     500	   2922043 ns/op	 1352504 B/op	   30747 allocs/op
```

This branch:

```
BenchmarkSnapshotCollectionByUUID-8    	  300000	      4492 ns/op	    1792 B/op	      39 allocs/op
BenchmarkSnapshotCollectionByName-8    	    1000	   1433058 ns/op	  533994 B/op	   14870 allocs/op
```

Reply:
I also had an idea to implement very simple "index" to lookup things by name (which is really frequent lookup): just a key in DB which is like `name` -> `UUID`. `ByName()` could use this index optimistically - if it's missing or doesn't point to right entry, `ByName()` falls back to full scan.

Reply:
This could be a way. I think we should see first how these changes are in terms of performance in a actual use case before we make the code even more complex.